pkg/asset/installconfig/aws/route53.go

```diff
@@ -148,37 +157,237 @@ func GetR53ClientCfg(sess *awss.Session, roleARN string) *aws.Config {
 }
 
 // CreateOrUpdateRecord Creates or Updates the Route53 Record for the cluster endpoint.
-func (c *Client) CreateOrUpdateRecord(ic *types.InstallConfig, target string, cfg *aws.Config) error {
-	zone, err := c.GetBaseDomain(ic.BaseDomain)
-	if err != nil {
-		return err
+func (c *Client) CreateOrUpdateRecord(ctx context.Context, ic *types.InstallConfig, target string, intTarget string, phzID string) error {
+	logrus.Debugf("Creating/Updating records for target=%s inTarget=%s in phzID=%s", target, intTarget, phzID)
+	useCNAME := cnameRegions.Has(ic.AWS.Region)
+	logrus.Debugf("Using CNAME: %v", useCNAME)
+	aliasZoneID := hostedZoneIDPerRegionNLBMap[ic.AWS.Region]
+	logrus.Debugf("aliasZone ID = %s", aliasZoneID)
+
+	logrus.Debugf("cluster domain: %s", ic.ClusterDomain())
+	apiName := fmt.Sprintf("api.%s.", ic.ClusterDomain())
+	logrus.Debugf("api name = %s", apiName)
+	apiIntName := fmt.Sprintf("api-int.%s.", ic.ClusterDomain())
+	logrus.Debugf("api-int name = %s", apiIntName)
+
+	// Create api record in public zone
+	if ic.Publish == types.ExternalPublishingStrategy {
+		logrus.Debugln("Public endpoints requested, creating public API record")
+		zone, err := c.GetBaseDomain(ic.BaseDomain)
+		if err != nil {
+			return err
+		}
+		logrus.Debugf("Public HZ ID: %s", aws.StringValue(zone.Id))
+
+		svc := route53.New(c.ssn) // we dont want to assume role here
+		if _, err := createRecord(ctx, svc, aws.StringValue(zone.Id), apiName, target, aliasZoneID, useCNAME); err != nil {
+			return fmt.Errorf("failed to create records for api (%s): %w", apiName, err)
+		}
+		logrus.Debugln("Created public API record in public zone")
+	} else {
+		logrus.Infoln("Public endpoints not requested, skipping public API record")
+	}
+
+	// Create service with assumed role for PHZ
+	svc := route53.New(c.ssn, GetR53ClientCfg(c.ssn, ic.AWS.HostedZoneRole))
+
+	// Create api record in private zone
+	if _, err := createRecord(ctx, svc, phzID, apiName, intTarget, aliasZoneID, useCNAME); err != nil {
+		return fmt.Errorf("failed to create records for api (%s): %w", apiName, err)
+	}
+	logrus.Debugln("Created public API record in private zone")
+
+	// Create api-int record in private zone
+	if _, err := createRecord(ctx, svc, phzID, apiIntName, intTarget, aliasZoneID, useCNAME); err != nil {
+		return fmt.Errorf("failed to create records for api-int (%s): %w", apiIntName, err)
 	}
+	logrus.Debugln("Created private API record in private zone")
 
-	params := &route53.ChangeResourceRecordSetsInput{
+	return nil
+}
+
+func createRecord(ctx context.Context, client *route53.Route53, zoneID, name, dnsName, aliasZoneID string, useCNAME bool) (*route53.ChangeInfo, error) {
+	recordSet := &route53.ResourceRecordSet{
+		Name: aws.String(name),
+	}
+	if useCNAME {
+		recordSet.SetType("CNAME")
+		recordSet.SetTTL(10)
```

Comment:
Considering CNAME will require a second DNS query to resolve the target name, is 10 seconds too short for this unchangeable record?
(and route53 bills by DNS query)

Reply:
That's the value we use in terraform as well. I don't mind increasing it if we think is safer.

Reply:
LGTM

Reply:
I believe we can keep the terraform for now, reviewing it later